clients/src/main/java/org/apache/kafka/clients/consumer/internals/StreamsMembershipManager.java

```diff
@@ -0,0 +1,823 @@
+                               final Map<String, SortedSet<Integer>> warmupTasks) {
+            this.localEpoch = localEpoch;
+            this.activeTasks = activeTasks;
+            this.standbyTasks = standbyTasks;
+            this.warmupTasks = warmupTasks;
+            if (localEpoch == NONE_EPOCH &&
+                (!activeTasks.isEmpty() || !standbyTasks.isEmpty() || !warmupTasks.isEmpty())) {
+                throw new IllegalArgumentException("Local epoch must be set if tasks are assigned.");
+            }
+        }
+
+        Optional<LocalAssignment> updateWith(final Map<String, SortedSet<Integer>> activeTasks,
+                                             final Map<String, SortedSet<Integer>> standbyTasks,
+                                             final Map<String, SortedSet<Integer>> warmupTasks) {
+            if (localEpoch != NONE_EPOCH) {
+                if (activeTasks.equals(this.activeTasks) &&
+                    standbyTasks.equals(this.standbyTasks) &&
+                    warmupTasks.equals(this.warmupTasks)) {
+
+                    return Optional.empty();
+                }
+            }
+
+            long nextLocalEpoch = localEpoch + 1;
+            return Optional.of(new LocalAssignment(nextLocalEpoch, activeTasks, standbyTasks, warmupTasks));
+        }
+
+        @Override
+        public String toString() {
+            return "LocalAssignment{" +
+                "localEpoch=" + localEpoch +
+                ", activeTasks=" + activeTasks +
+                ", standbyTasks=" + standbyTasks +
+                ", warmupTasks=" + warmupTasks +
+                '}';
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            LocalAssignment that = (LocalAssignment) o;
+            return localEpoch == that.localEpoch &&
+                Objects.equals(activeTasks, that.activeTasks) &&
+                Objects.equals(standbyTasks, that.standbyTasks) &&
+                Objects.equals(warmupTasks, that.warmupTasks);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(localEpoch, activeTasks, standbyTasks, warmupTasks);
+        }
+    }
+
+    static final Utils.TopicPartitionComparator TOPIC_PARTITION_COMPARATOR = new Utils.TopicPartitionComparator();
+
+    private final Logger log;
+
+    private final StreamsAssignmentInterface streamsAssignmentInterface;
+
+    private final SubscriptionState subscriptionState;
+
+    private final ConsumerMetadata consumerMetadata;
+
+    private final BackgroundEventHandler backgroundEventHandler;
+
+    private final Map<String, Uuid> assignedTopicIdCache = new HashMap<>();
+
+    private MemberState state;
+
+    private final String groupId;
+
+    private String memberId = "";
+
+    private final Optional<String> groupInstanceId = Optional.empty();
```

Comment:
General point. It seems like you copied over a lot of code, but you did not copy the javadoc comments. Why?

Reply:
Yeah, I know. I did not want to blindly copy over the javadocs, but wanted to write new ones to get them right. I still want to write them, just haven't had time yet.

Reply:
I added some javadocs. Can we postpone the remaining javadocs to when this class stabilizes and we will extract shared code? I believe that makes more sense.